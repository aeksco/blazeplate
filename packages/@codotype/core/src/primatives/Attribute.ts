import { v4 as uuidv4 } from "uuid";
import { TokenCasing, Datatype, AddonsValue } from "..";
import { CreatedBy, CreatedByValues } from "../created-by";
import { Attribute } from "../attribute";

// // // //

interface AttributeBuilderParams {
    id?: string;
    identifiers: TokenCasing;
    datatype: Datatype;
    defaultValue?: null | string | boolean | number;
    internalNote?: string;
    source?: CreatedBy;
    locked?: boolean;
    addons?: AddonsValue;
}

export class AttributeBuilder implements Attribute {
    id: string = uuidv4();
    identifiers: TokenCasing;
    datatype: Datatype;
    defaultValue: string | number | boolean = null;
    internalNote: string = "";
    createdBy: CreatedBy = CreatedByValues.user;
    locked: boolean = false;
    addons: AddonsValue = {};

    constructor(params: AttributeBuilderParams) {
        this.identifiers = params.identifiers;
        this.datatype = params.datatype;

        this.id = params.id || this.id;
        this.defaultValue = params.defaultValue || this.defaultValue;
        this.internalNote = params.internalNote || this.internalNote;
        this.createdBy = params.source || this.createdBy;
        this.locked = params.locked || this.locked;
        this.addons = params.addons || this.addons;
    }
}
