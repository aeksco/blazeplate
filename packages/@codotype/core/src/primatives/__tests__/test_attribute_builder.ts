import { AttributeBuilder } from "../Attribute";
import { Datatype } from "../..";

// Mocks UUID module
jest.mock("uuid", () => ({
    __esModule: true,
    v4: () => "1234-5678-1234-5678",
}));

// // // //

describe("AttributeBuilder", () => {
    test("works", () => {
        const attribute = new AttributeBuilder({
            datatype: Datatype.STRING,
            identifiers: {
                title: "My String",
                snake: "my_string",
                camel: "myString",
                pascal: "MyString",
                kebab: "my-string",
            },
        });

        expect(attribute).toMatchSnapshot();
    });
});
