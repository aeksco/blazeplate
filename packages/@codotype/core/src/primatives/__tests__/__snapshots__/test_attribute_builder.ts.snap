// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttributeBuilder works 1`] = `
AttributeBuilder {
  "addons": Object {},
  "datatype": "STRING",
  "defaultValue": null,
  "id": "1234-5678-1234-5678",
  "identifiers": Object {
    "camel": "myString",
    "kebab": "my-string",
    "pascal": "MyString",
    "snake": "my_string",
    "title": "My String",
  },
  "internalNote": "",
  "locked": false,
  "source": "user",
}
`;
